module/zfs/zfs_znode.c

```diff
@@ -1049,6 +1075,11 @@ zfs_xvattr_set(znode_t *zp, xvattr_t *xvap, dmu_tx_t *tx)
 		    zp->z_pflags, tx);
 		XVA_SET_RTN(xvap, XAT_SPARSE);
 	}
+	if (XVA_ISSET_REQ(xvap, XAT_PROJINHERIT)) {
+		ZFS_ATTR_SET(zp, ZFS_PROJINHERIT, xoap->xoa_projinherit,
+		    zp->z_pflags, tx);
```

Comment:
@Nasf-Fan @ahrens I think it would be best to adopt the semantics already established by xfs for project quotas.  Users will have existing expectations about the behavior.  Plus my understanding is there is a desire that this become generic functionality which is why it has already been added to ext4.   Having zfs behave differently would be confusing.

The good news is it seems to me they've made some reasonable decisions about how directory trees should be handled.  See the [Directory Tree Quota](https://linux.die.net/man/8/xfs_quota) section of `xfs_quota(8)`.  The expected behavior adapted for zfs and summarized would be:

* A managed tree must be setup initially using the __zfs project -s__ command.
  * Recursively sets an inode flag and the project identifier on every file in the affected tree.
  * New files created in the tree will be automatically be accounted to the tree.
  * Hard links to a file in the tree will only succeed if the project identifier matches the project identifier for the tree.
  * As a privileged user __zfs project__ can be used to set the project ID for an arbitrary file.
* The __zfs project -C__ option clears a tree from project quota control.
* The __zfs project -c__ option can be used to check whether a tree is setup.

There are two optional configuration files which should be supported `/etc/projects` and `/etc/projid`.

https://linux.die.net/man/5/projects
https://linux.die.net/man/5/projid